NMS-13807: Phase 1 flows documentation: "Basic" setup (#3966)
OpenNMS/opennms: docs/modules/operation/pages/flows/basic.adoc

[[flows-basic]]

> = Start here: basic flows setup

Author: No worries. Happy to change it.

This section describes how to get started with flows in a basic setup to collect, enrich (classify), persist, and visualize flows.

== Requirements

Make sure you have the following before you set up flows:

* OpenNMS up and running.
* Device(s) that sends flows visible to OpenNMS and monitored with SNMP.
* Elasticsearch cluster set up with the link:https://github.com/OpenNMS/elasticsearch-drift-plugin[Elasticsearch Drift plugin] installed on every Elasticsearch node.
** The Drift plugin persists and queries flows that {page-component-title} collects.
The Drift version must match the targeted Elasticsearch version.

> ** (optional) Configure Elasticsearch variables like `search.max_buckets` if the default values are not sufficient.

Author: That's a good question. @indigo423 , any suggestions?

> ** (optional) Configure Elasticsearch variables like `search.max_buckets` if the default values are not sufficient.

Reviewer: I don't have a list. Sorry.

** (optional) Create a job to clean the jobs so that the disk does not fill up; for example, keep X days of flows.
Filled disks are a challenging problem to address for non-Elasticsearch experts.
We recommend the Elasticsearch https://www.elastic.co/guide/en/elasticsearch/client/curator/current/index.html[Curator tool] to do this.
** Monitor the Elasticsearch stack in OpenNMS to get an alarm if it goes down.
* Set up xref:helm/ROOT/index.adoc#welcome-index[OpenNMS Helm] with Grafana (to visualize flows)

> ** Configure the xref:getting_started/basic_walkthrough.adoc#bw-ds-setup[flow and performance data sources]

Author: Yes, I see they made the build fail. Will update.

> ** Configure the xref:getting_started/basic_walkthrough.adoc#bw-ds-setup[flow and performance data sources]

Author: @indigo423 or @mmahacek ... I am not sure what I am missing in the HELM links. I know they are in a separate repo, so is it even possible to create links to those docs?

> ** Configure the xref:getting_started/basic_walkthrough.adoc#bw-ds-setup[flow and performance data sources]

Reviewer 2: For some reason I didn't think we could do cross-product xref's like this. I thought they had to be external links.

> ** Configure the xref:getting_started/basic_walkthrough.adoc#bw-ds-setup[flow and performance data sources]

Author: I think you are correct. Once we have the floating "latest" tag, we can effectively use external links. I'll remove these for now.

== Configure OpenNMS to communicate with Elasticsearch

OpenNMS needs to talk to Elasticsearch and know where to store the flows data it collects (persistence).

> From a Karaf shell on your {page-component-title} instance, update `$OPENNMS_HOME/etc/org.opennms.features.flows.persistence.elastic.cfg` to configure the flow persistence to use your Elasticsearch cluster:

Reviewer 2: ```suggestion
From a Karaf shell on your {page-component-title} instance, update `$\{OPENNMS_HOME}/etc/org.opennms.features.flows.persistence.elastic.cfg` to configure the flow persistence to use your Elasticsearch cluster:
```

> [source, console]
> ----
> $ ssh -p 8101 admin@localhost
> ...
> admin@opennms()> config:edit org.opennms.features.flows.persistence.elastic
> admin@opennms()> config:property-set elasticUrl http://elastic:9200
> admin@opennms()> config:update
> ----

Reviewer 2: ```suggestion
.Connect to Karaf shell
[source, console]
----
ssh -p 8101 admin@localhost
----

.Configure Elasticsearch settings within Karaf
[source, karaf]
----
config:edit org.opennms.features.flows.persistence.elastic
config:property-set elasticUrl http://elastic:9200
config:update
----
```

We also recommend setting the following:

> [source, xml]

Reviewer 2: ```suggestion
.Edit (or create) `$\{OPENNMS_HOME}/etc/org.opennms.features.flows.persistence.elastic.cfg`
[source, xml]
```

----
# ElasticSearch persistence configuration

> elasticUrl = http://10.10.3.218:9200

Reviewer 2: ```suggestion
elasticUrl = http://10.10.3.218:9200 <1>
```

connTimeout = 30000
readTimeout = 300000
settings.index.number_of_replicas = 0
settings.index.number_of_shards=1
settings.index.refresh_interval=10s
elasticIndexStrategy=daily
----

> 

Reviewer 2: ```suggestion
<1> Replace with comma-separated list of Elasticsearch nodes

```

> See <<elasticsearch/introduction.adoc#ga-elasticsearch-integration-configuration, General Elasticsearch Configuration>> for a complete set of options.

Reviewer 2: ```suggestion
NOTE: See <<elasticsearch/introduction.adoc#ga-elasticsearch-integration-configuration, General Elasticsearch Configuration>> for a complete set of options.
```

Author: NH style guide prefers to limit the use of notes, and since this is a separate paragraph with a link, I think it already stands out.

== Enable protocols

> Update `$OPENNMS_HOME/etc/telemetryd-configuration.xml` to enable one or more of the protocols you want to handle.

Reviewer 2: ```suggestion
Update `$\{OPENNMS_HOME}/etc/telemetryd-configuration.xml` to enable one or more of the protocols you want to handle.
```

This example enables the NetFlow v5 protocol.
Use the same process for any of the other flow-related protocols.

[source, xml]
----
<listener name="Netflow-5-UDP-8877" class-name="org.opennms.netmgt.telemetry.listeners.UdpListener" enabled="true">
    <parameter key="port" value="8877"/>

    <parser name="Netflow-5-Parser" class-name="org.opennms.netmgt.telemetry.protocols.netflow.parser.Netflow5UdpParser" queue="Netflow-5" />
</listener>

<queue name="Netflow-5">
    <adapter name="Netflow-5-Adapter" class-name="org.opennms.netmgt.telemetry.protocols.netflow.adapter.netflow5.Netflow5Adapter" enabled="true">
    </adapter>
</queue>
----

Send a `reloadDaemonConfig` event via the CLI to apply the changes without restarting:

[source, console]
----

> $OPENNMS_HOME/bin/send-event.pl -p 'daemonName Telemetryd' uei.opennms.org/internal/reloadDaemonConfig

Reviewer 2: ```suggestion
$\{OPENNMS_HOME}/bin/send-event.pl -p 'daemonName Telemetryd' uei.opennms.org/internal/reloadDaemonConfig
```

----

This opens a UDP socket bound to `0.0.0.0:8877` to which NetFlow v5 messages are forwarded.

=== Multi-port listener

If you are monitoring multiple flow protocols, you normally need to set up a flow listener for each one, on its own UDP port.

> By default, OpenNMS enables its multi-port listener option, which monitors multiple protocols on a single UDP port (9999).

Reviewer 2: ```suggestion
By default, {page-component-title} enables a multi-port listener option, which monitors multiple protocols on a single UDP port (9999).
```

> If required, edit `etc/telemetryd-configuration.xml` to change the port number or add/remove protocols.

Reviewer 2: ```suggestion
If desired, edit `$\{OPENNMS_HOME}/etc/telemetryd-configuration.xml` to change the port number or add/remove protocols.
```

> 

Reviewer 2: ```suggestion

IMPORTANT: Make sure any ports you configure for receiving flow data are added to your firewall allow list.

```

== Enable flows on your device(s)

Configure your devices to send flows.
Refer to the manufacturer's documentation.
You will need to set up the flow receiver, which is OpenNMS {page-component-title}, and enable sending flows per interface on the firewall.

== Link the web UI to Helm

To access flow-related graphs from the {page-component-title} web interface, you must configure a link to your instance of OpenNMS Helm.

> ----
> $ ssh -p 8101 admin@localhost
> ...
> admin@opennms()> config:edit org.opennms.netmgt.flows.rest
> admin@opennms()> config:property-set flowGraphUrl 'http://grafana:3000/dashboard/flows?node=$nodeId&interface=$ifIndex'
> admin@opennms()> config:update
> ----

Reviewer 2: ```suggestion
.Connect to Karaf shell
[source, console]
----
ssh -p 8101 admin@localhost
----

.Configure Helm settings within Karaf
[source, karaf]
----
config:edit org.opennms.netmgt.flows.rest
config:property-set flowGraphUrl 'http://grafana:3000/dashboard/flows?node=$nodeId&interface=$ifIndex'
config:update
----
```

settings.index.refresh_interval=10s
elasticIndexStrategy=daily
----
<1> Replace with comma-separated list of Elasticsearch nodes.

See <<elasticsearch/introduction.adoc#ga-elasticsearch-integration-configuration, General Elasticsearch Configuration>> for a complete set of options.

== Enable protocols

Update `$\{OPENNMS_HOME}/etc/telemetryd-configuration.xml` to enable one or more of the protocols you want to handle.

This example enables the NetFlow v5 protocol.
Use the same process for any of the other flow-related protocols.

[source, xml]
----
<listener name="Netflow-5-UDP-8877" class-name="org.opennms.netmgt.telemetry.listeners.UdpListener" enabled="true">
    <parameter key="port" value="8877"/>

    <parser name="Netflow-5-Parser" class-name="org.opennms.netmgt.telemetry.protocols.netflow.parser.Netflow5UdpParser" queue="Netflow-5" />
</listener>

<queue name="Netflow-5">
    <adapter name="Netflow-5-Adapter" class-name="org.opennms.netmgt.telemetry.protocols.netflow.adapter.netflow5.Netflow5Adapter" enabled="true">
    </adapter>
</queue>
----

Send a `reloadDaemonConfig` event via the CLI to apply the changes without restarting:

[source, console]
----

> $\{OPENNMS_HOME}/bin/send-event.pl -p 'daemonName Telemetryd' uei.opennms.org/internal/reloadDaemonConfig

Reviewer: @Bonrob2 Is that still the way we want to offer? 

@mmahacek @indigo423 I've never tried it but can't we do something like:

`ssh -p 8101 admin@localhost 'reload-daemon telemetryd'`

----

This opens a UDP socket bound to `0.0.0.0:8877` to which NetFlow v5 messages are forwarded.

=== Multi-port listener

If you are monitoring multiple flow protocols, you normally need to set up a flow listener for each one, on its own UDP port.

By default, {page-component-title} enables a multi-port listener option, which monitors multiple protocols on a single UDP port (9999).
If desired, edit `$\{OPENNMS_HOME}/etc/telemetryd-configuration.xml` to change the port number or add/remove protocols.

IMPORTANT: Make sure any ports you configure for receiving flow data are added to your firewall allow list.

== Enable flows on your device(s)

Configure your devices to send flows.
Refer to the manufacturer's documentation.

> You will need to set up the flow receiver, which is OpenNMS {page-component-title}, and enable sending flows per interface on the firewall.

Reviewer: ```suggestion
You will need to set up the flow receiver, which is OpenNMS {page-component-title}, and probably enable sending flows per interface on the firewall.
```
@Bonrob2 I guess this is from my Fortinet example. Is that procedure the same on other devices? Never had to set up another manufacturer.

Author: I'm not sure if it is the same on other devices. We can modify this to indicate that it is specific to Fortinet.

== Link the web UI to Helm

To access flow-related graphs from the {page-component-title} web interface, you must configure a link to your instance of OpenNMS Helm.

.Connect to Karaf shell
[source, console]
----
ssh -p 8101 admin@localhost
----

.Configure Helm settings within Karaf
[source, karaf]
----
config:edit org.opennms.netmgt.flows.rest
config:property-set flowGraphUrl 'http://grafana:3000/dashboard/flows?node=$nodeId&interface=$ifIndex'
config:update
----

> NOTE: This URL can optionally point to other tools as well.

Reviewer: I am wondering which other tools are meant here. Maybe add 1-2 examples?

Author: Great idea. @mmahacek or @indigo423 any thoughts on other tools?

Reviewer 2: I see the potential, but I don't know of any other tools that are feasible at this time.

[[flows-basic]]
= Basic Flows Setup

This section describes how to get started with flows to collect, enrich (classify), persist, and visualize flows.

== Requirements

Make sure you have the following before you set up flows:

* OpenNMS up and running.
* Device(s) that sends flows visible to OpenNMS and monitored with SNMP.
* Elasticsearch cluster set up with the link:https://github.com/OpenNMS/elasticsearch-drift-plugin[Elasticsearch Drift plugin] installed on every Elasticsearch node.
** The Drift plugin persists and queries flows that {page-component-title} collects.
The Drift version must match the targeted Elasticsearch version.
** (optional) Configure Elasticsearch variables like `search.max_buckets` or maximum heap size `ES-JAVA_OPTS`if the default values are not sufficient for your volume of flows or number of nodes.
** (optional) Create a job to clean the indices so that the disk does not fill up; for example, keep X days of flows.
Filled disks are a challenging problem to address for non-Elasticsearch experts.

> We recommend the Elasticsearch https://www.elastic.co/guide/en/elasticsearch/client/curator/current/index.html[Curator tool] to do this.

Reviewer 2: ```suggestion
We recommend the Elasticsearch link:https://www.elastic.co/guide/en/elasticsearch/client/curator/current/index.html[Curator tool] to do this.
```
